### ISSUE TITLE
[ESPnet2] Resume training from best (valid) epoch instead of last epoch

### ISSUE THREAD

**Issue by yash-s20** [espnet/espnet#2645]
Hi
I'm trying to manipulate the number of training epochs to achieve training the same network with multiple datasets one-by-one.
So I swap training sets, increase number of epochs in the config file, and start again.

What I've noticed is that my accuracy at the end of this ordeal might improve if I pick the best accuracy epoch and continue training with a new dataset, rather than the last epoch.

If I change the soft link of the last epoch to the best one, would it perform the way I intend it to? Otherwise, is there a flag to resume training from the best epoch?
Or do you suggest another approach for what I'm trying to achieve?

Thanks

**Comment by kamo-naoyuki** [comment]
> If I change the soft link of the last epoch to the best one, would it perform the way I intend it to? Otherwise, is there a flag to resume training from the best epoch?

Now it's impossible to resume the state of the best epoch because it's not saved.
When resuming the following states are loaded,

- Model
- Optimizer
- Scheduler
- Reporter
- amp Scaler

These states are saved in checkpoint.pth for the last epoch, but the model state is only saved for each epoch.

However, You can restart the training with the pretrained model instead using `--pretrained_key` and `--pretrained_path`.

**Comment by yash-s20** [comment]
Thanks for the reply.
I now understand how the last model works. It picks up the checkpoint rather than the last model.

I saw the implementation of ``pretrain_key`` and ``pretrain_path`` but never got around to use it, thought this was easier. Anyway, as I understand, if I want to load the entire model, ``pretrain_key`` should ``"none"`` right?

**Comment by kamo-naoyuki** [comment]
Anyway, as I understand, if I want to load the entire model, `pretrain_key` should `"none"` right?

Yes
